src/daemon/daemon.cpp

```diff
@@ -2347,6 +2501,145 @@ catch (const std::exception& e)
     status_promise->set_value(grpc::Status(grpc::StatusCode::INTERNAL, e.what(), ""));
 }
 
+void mp::Daemon::snapshot(const mp::SnapshotRequest* request,
+                          grpc::ServerReaderWriterInterface<SnapshotReply, SnapshotRequest>* server,
+                          std::promise<grpc::Status>* status_promise)
+try
+{
+    mpl::ClientLogger<SnapshotReply, SnapshotRequest> logger{mpl::level_from(request->verbosity_level()),
+                                                             *config->logger,
+                                                             server};
+
+    const auto& instance_name = request->instance();
+    auto [instance_trail, status] = find_instance_and_react(operative_instances,
+                                                            deleted_instances,
+                                                            instance_name,
+                                                            require_operative_instances_reaction);
+
+    if (status.ok())
+    {
+        assert(instance_trail.index() == 0);
+        auto* vm_ptr = std::get<0>(instance_trail)->second.get();
+        assert(vm_ptr);
+
+        using St = VirtualMachine::State;
+        if (auto state = vm_ptr->current_state(); state != St::off && state != St::stopped)
+            return status_promise->set_value(
+                grpc::Status{grpc::INVALID_ARGUMENT, "Multipass can only take snapshots of stopped instances."});
+
+        auto snapshot_name = request->snapshot();
+        if (!snapshot_name.empty() && !mp::utils::valid_hostname(snapshot_name))
+            return status_promise->set_value(
+                grpc::Status{grpc::INVALID_ARGUMENT, fmt::format(R"(Invalid snapshot name: "{}".)", snapshot_name)});
+
+        const auto spec_it = vm_instance_specs.find(instance_name);
+        assert(spec_it != vm_instance_specs.end() && "missing instance specs");
+
+        SnapshotReply reply;
+
+        {
+            const auto snapshot = vm_ptr->take_snapshot(spec_it->second, snapshot_name, request->comment());
+
+            reply.set_snapshot(snapshot->get_name());
+        }
```

Comment:
Curious why this is in its own scope?

Reply:
Hmm, `snapshot` is a `shared_ptr` there, so I think this was meant to avoid prolonging, beyond what is strictly necessary, the lifetime of a snapshot that might otherwise have already been deleted by another thread. But I guess this was more important at the beginning of development, when I was using locked objects for thread-safety (to limit the time that locks were held.) 

I guess I can remove it and let the snapshot live for a little longer while the daemon replies to the client.

Reply:
Actually, I'll just use the snapshot as an rvalue here and let it be destroyed right after.